### PR TITLE
Robotics (ROS 2) project template

### PR DESCRIPTION
**Pull request by adamdbrw**
This PR adds a robotics project template and makes use of the Warehouse and ROSBot XL Gems which are added through #75.
It is intended to be a good starting project for ROS developers interested in prototyping with O3DE.  

- [x] Add description on what it does here.
- [x] Clean the PR up (no unnecessary files, review additional files such as LFS filter)

### CODE REVIEW

File: Templates/Ros2ProjectTemplate/Template/Config/shader_global_build_options.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+    "Type": "JsonSerialization",
+    "Version": 1,
+    "ClassName": "GlobalBuildOptions",
+    "ClassData": {
+        "ShaderCompilerArguments" : {
+            "DefaultMatrixOrder" : "Row",
+            "AzslcAdditionalFreeArguments" : "--strip-unused-srgs"
+        }
+    }
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Config/shader_global_build_options.json:11]: Add newline at end of file

File: Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/config/slam_params.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    # Scan Matcher Parameters
+    distance_variance_penalty: 0.5      
+    angle_variance_penalty: 1.0    
+
+    fine_search_angle_offset: 0.00349     
+    coarse_search_angle_offset: 0.349   
+    coarse_angle_resolution: 0.0349        
+    minimum_angle_penalty: 0.9
+    minimum_distance_penalty: 0.5
+    use_response_expansion: true
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/config/slam_params.yaml:73]: Add newline at end of file

File: Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/navigation.launch.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        Node(
+            package='rviz2',
+            executable='rviz2',
+            name='slam',
+            output='screen',
+            arguments=[
+                '-d', str(pathlib.Path(__file__).parent.absolute().joinpath('config', 'config.rviz'))
+            ]
+        ),
+    ])
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/navigation.launch.py:44]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/slam.launch.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            parameters=[{
+                'min_height': 0.0,
+                'max_height': 0.4,
+                'range_min': 0.05
+            }],
+            remappings=[
+                ('/cloud_in', '/pc'),
+            ]
+        )
+    ])
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/slam.launch.py:47]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Gem/gem.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+    "gem_name": "${Name}",
+    "display_name": "${Name}",
+    "license": "What license ${Name} uses goes here: i.e. Apache-2.0 or MIT",
+    "license_url": "Link to the license web site goes here: i.e. https://opensource.org/licenses/Apache-2.0 Or https://opensource.org/licenses/MIT",
+    "origin": "The name of the originator goes here. i.e. XYZ Inc.",
+    "origin_url": "The primary repo for ${Name} goes here: i.e. http://www.mydomain.com",
+    "type": "",
+    "summary": "A short description of ${Name}.",
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Gem/gem.json:9]: Fill in boilerplate with specific inforamtion

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab [hunk 6]
```diff
@@ -0,0 +1,2312 @@
...
+                },
+                {
+                    "op": "replace",
+                    "path": "/Entities/Entity_[394238657263]/Components/Component_[12399796004250012151]/Configuration/Inertia tensor/8",
+                    "value": 1.0
+                }
+            ]
+        }
+    }
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab:2312]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Platform/Android/android_project.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+    "Tags": ["Android"],
+    "android_settings" : {
+        "package_name" : "org.o3de.${Name}",
+        "version_number" : 1,
+        "version_name" : "1.0.0.0",
+        "orientation" : "landscape"
+    }
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Platform/Android/android_project.json:9]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Platform/Linux/linux_project.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+  "Tags": ["Linux"]
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Platform/Linux/linux_project.json:3]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Platform/Mac/mac_project.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
+  "Tags": ["Mac"]
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Platform/Mac/mac_project.json:3]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Platform/Windows/windows_project.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+  "Tags": ["Windows"]
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Platform/Windows/windows_project.json:3]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Platform/iOS/ios_project.json [hunk 11]
```diff
@@ -0,0 +1,3 @@
+{
+    "Tags": ["iOS"]
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Platform/iOS/ios_project.json:3]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Registry/physxsystemconfiguration.setreg [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+                                    "ReadOnly": true
+                                }
+                            ]
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Registry/physxsystemconfiguration.setreg:98]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Resources/Platform/Mac/Images.xcassets/Contents.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Resources/Platform/Mac/Images.xcassets/Contents.json:6]: newline at end

File: Templates/Ros2ProjectTemplate/Template/Resources/Platform/iOS/Images.xcassets/AppIcon.appiconset/Contents.json [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+      "idiom" : "ios-marketing",
+      "filename" : "iOSAppStoreIcon1024x1024.png",
+      "scale" : "1x"
+    }
+  ],
+  "info" : {
+    "version" : 1,
+    "author" : "xcode"
+  }
+}
```
Comment (byrcolin) [inline Templates/Ros2ProjectTemplate/Template/Resources/Platform/iOS/Images.xcassets/AppIcon.appiconset/Contents.json:116]: newline at end

**Review by byrcolin — APPROVED** [review]
just nit fixes: newlines at end of some files, otherwise looks ok.

**Comment by michalpelka** [comment]
Rebased and squashed (old Pawel's branch is here: https://github.com/RobotecAI/o3de-extras/tree/ros2_project_template--20230111_1723)
It is rebased on top of other PRs: #75 and #86 .
Thank you for reviewing @byrcolin I updated the code, since I am not an author I cannot resolve conversations.

**Comment by michalpelka** [comment]
Addressed previous comments, tested, and rebased on top of #75.

File: Templates/Ros2ProjectTemplate/README.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+```shell
+source /opt/ros/humble/setup.bash
+./build/linux/bin/profile/o3de
+```
+
+### 1. Install ROS2 packages
+
+```shell
+sudo apt install ros-${ROS_DISTRO}-slam-toolbox ros-${ROS_DISTRO}-navigation2 ros-${ROS_DISTRO}-nav2-bringup ros-${ROS_DISTRO}-pointcloud-to-laserscan ros-${ROS_DISTRO}-teleop-twist-keyboard ros-${ROS_DISTRO}-ackermann-msgs ros-${ROS_DISTRO}-gazebo-msgs ros-${ROS_DISTRO}-control-toolbox
```
[the remark this answers is not in the document]
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/README.md]: Done

File: Templates/Ros2ProjectTemplate/Template/Config/shader_global_build_options.json (hunk 1 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Config/shader_global_build_options.json]: Done

File: Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/config/slam_params.yaml (hunk 2 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/config/slam_params.yaml]: Done

File: Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/slam.launch.py (hunk 4 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Examples/slam_navigation/launch/slam.launch.py]: Done

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab (hunk 6 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab]: Done

File: Templates/Ros2ProjectTemplate/Template/Platform/Android/android_project.json (hunk 7 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Platform/Android/android_project.json]: Done

File: Templates/Ros2ProjectTemplate/Template/Platform/Mac/mac_project.json (hunk 9 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Platform/Mac/mac_project.json]: Done

File: Templates/Ros2ProjectTemplate/Template/Platform/iOS/ios_project.json (hunk 11 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Platform/iOS/ios_project.json]: Done

File: Templates/Ros2ProjectTemplate/Template/Registry/physxsystemconfiguration.setreg (hunk 12 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Registry/physxsystemconfiguration.setreg]: Done

File: Templates/Ros2ProjectTemplate/Template/Resources/Platform/Mac/Images.xcassets/Contents.json (hunk 13 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Resources/Platform/Mac/Images.xcassets/Contents.json]: Done

File: Templates/Ros2ProjectTemplate/Template/Resources/Platform/iOS/Images.xcassets/AppIcon.appiconset/Contents.json (hunk 14 above)
Comment (michalpelka) [reply Templates/Ros2ProjectTemplate/Template/Resources/Platform/iOS/Images.xcassets/AppIcon.appiconset/Contents.json]: Done

**Review by hultonha — APPROVED** [review]

**Pull request merged**